[tune] Add `set_max_concurrency` to Searcher API

Yard1 opened the pull request:
<!-- Thank you for your contribution! Please review https://github.com/ray-project/ray/blob/master/CONTRIBUTING.rst before opening a pull request. -->

<!-- Please add a reviewer to the assignee section when you create a PR. If you don't have the access to it, we will shortly find a reviewer and assign them to your PR. -->

## Why are these changes needed?

Adds a `set_max_concurrency` method to the Searcher API. This method allows for the `ConcurrencyLimiter` to override the `max_concurrency` value on searchers with custom internal logic for limiting concurrency (atm. SigOpt and HEBO). This PR also changes the initialisation of SigOpt and HEBO optimisers to happen on `set_search_properties` instead of in the constructor, so that the new `max_concurrency` is respected.

Furthermore, this PR breaks up `test_tune_restore.py` into `test_tune_restore_warm_start.py` and `test_tune_restore.py` to deal with a timeout, and ensures that the automatic application of `ConcurrencyLimiter` in `tune.run` doesn't override a user-defined `ConcurrencyLimiter`.

<!-- Please give a short summary of the change and the problem this solves. -->

## Related issue number

<!-- For example: "Closes #1234" -->

## Checks

- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [x] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [x] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [x] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

Yard1 commented on python/ray/tune/suggest/suggestion.py:
```diff
@@ -382,7 +406,25 @@ def __init__(self,
         super(ConcurrencyLimiter, self).__init__(
             metric=self.searcher.metric, mode=self.searcher.mode)
 
+    def set_max_concurrency(self, max_concurrent: int) -> bool:
+        # Determine if this behavior is acceptable, or if it should
+        # raise an exception.
+        self.max_concurrent = max_concurrent
+        return True
```
Feedback here would be appreciated @krfricke @xwjiang2010

xwjiang2010 replied:
I like it!

xwjiang2010 commented on python/ray/tune/suggest/suggestion.py:
```diff
@@ -372,17 +395,41 @@ def __init__(self,
+    def set_search_properties(self, metric: Optional[str], mode: Optional[str],
+                              config: Dict, **spec) -> bool:
+        self._set_searcher_max_concurrency()
+        return set_search_properties_backwards_compatible(
```
What is our deprecation cycle? I feel like we can safely remove this already?

xwjiang2010 replied:
Actually I have just removed it myself :)

Yard1 commented on python/ray/tune/suggest/hebo.py:
```diff
@@ -166,13 +171,19 @@ def __init__(
         self._live_trial_mapping = {}
 
         self._max_concurrent = max_concurrent
+        self._setup_optimizer_again = False
```
```suggestion
        self._max_concurrent = max_concurrent
```

Yard1 commented:
@xwjiang2010 @krfricke I have updated the PR. I had to break up the `tune_test_restore` into two files, as HEBO was taking a lot longer with concurrency correctly limited. Before, it was using its default mode of 8 concurrent trials despite the test using only one CPU - it didn't impact the test and the side effect of that was that it cached the suggestions, speeding it up a lot. I had fixed that but I left the test broken up as I feel that it makes more sense now, but let me know if you'd rather have it back in one file.

krfricke commented on python/ray/tune/suggest/ax.py:
```diff
@@ -258,9 +258,9 @@ def suggest(self, trial_id: str) -> Optional[Dict]:
-        if self.max_concurrent:
-            if len(self._live_trial_mapping) >= self.max_concurrent:
-                return None
+        # if self.max_concurrent:
+        #     if len(self._live_trial_mapping) >= self.max_concurrent:
+        #         return None
```
Remove?

krfricke commented on python/ray/tune/suggest/bayesopt.py:
```diff
@@ -245,10 +245,10 @@ def suggest(self, trial_id: str) -> Optional[Dict]:
-        if self.max_concurrent and self.max_concurrent <= total_live_trials:
-            # we stop the suggestion and return None.
-            return None
+        # total_live_trials = len(self._live_trial_mapping)
+        # if self.max_concurrent and self.max_concurrent <= total_live_trials:
+        #     # we stop the suggestion and return None.
+        #     return None
```
Remove?

krfricke commented on python/ray/tune/suggest/hyperopt.py:
```diff
@@ -265,9 +265,9 @@ def suggest(self, trial_id: str) -> Optional[Dict]:
-        if self.max_concurrent:
-            if len(self._live_trial_mapping) >= self.max_concurrent:
-                return None
+        # if self.max_concurrent:
+        #     if len(self._live_trial_mapping) >= self.max_concurrent:
+        #         return None
```
Remove?

krfricke commented on python/ray/tune/suggest/nevergrad.py:
```diff
@@ -240,9 +240,9 @@ def suggest(self, trial_id: str) -> Optional[Dict]:
-        if self.max_concurrent:
-            if len(self._live_trial_mapping) >= self.max_concurrent:
-                return None
+        # if self.max_concurrent:
+        #     if len(self._live_trial_mapping) >= self.max_concurrent:
+        #         return None
```
Remove?

xwjiang2010 reviewed (approved):
Thanks Antoni! Yes, let's split up the tests!!

krfricke commented on python/ray/tune/suggest/hebo.py:
```diff
@@ -219,11 +228,12 @@ def set_search_properties(self, metric: Optional[str], mode: Optional[str],
+        if not self._metric:
             self._metric = metric
-        if mode:
+        if not self._mode:
```
For all other searchers, we have `if mode:` here, in order to always overwrite the mode if passed to `tune.run()`. Should we keep this here, too?

Yard1 replied:
You are right, it should be consistent. Changed

krfricke reviewed (commented):
Looks good, some question about commented out code paths.
